Initial obstacle avoidance implementations version 1.0.0

drakeaxelrod opened the pull request:
Obstacle avoidance using infrared sensor and ultrasonic sensor.
Emergency breaks that kick in at predefined distances. refactored
the Publisher class to a method that returns a already logged in
and connected mqtt client.

drakeaxelrod committed:
Add project structure & basic mqtt setup

 Created src for the python portion of the project.
 Added a wrapper class for mqtt to make it easier to use.
 Created .env with python mqtt secrets, and secrets.hpp
 for ino mqtt secrets. amended gitignore to include .env
 and secrets.hpp

drakeaxelrod committed:
Add arrow key controls via mqtt messages

cleaned up some comments in the code that shouldn't have been there.
Added arrow key controls (functions the same as demo car). The arrow
keys send mqtt messages so the car can be controled from anywhere
if you running the program.

drakeaxelrod committed:
Initial obstacle avoidance implementations

Obstacle avoidance using infrared sensor and ultrasonic sensor.
Emergency breaks that kick in at predefined distances. refactored
the Publisher class to a method that returns a already logged in
and connected mqtt client.

VernitaJ reviewed (dismissed):
Basic setup complete with MQTT implemented. smartcar.ino to be updated with obstacle avoidance code implemented 07/04/2021.

platisd reviewed (commented):
Where's the obstacle avoidance implemented? Also, you should gitignore binary files!

platisd reviewed (commented):
OK, i've seen you've removed the binaries if you had some (i may be confusing you with another group).
That being said, I don't think you need that `.lock` file

VernitaJ commented:
> Where's the obstacle avoidance implemented? Also, you should gitignore binary files!

The smartcar.ino file wasn't updated in this branch, which means the obstacle avoidance isn't actually represented here at all at the moment. We caught this error in our meeting this afternoon and will update soonest.

MrJonte commented:
The obstacle avoidance is actually implemented in the keyboard class. Is this a hint that you want us to comment in the code that this is the obstacle avoidance? 
And/Or do you want us to implement the obstacle avoidance within the .ino file?

platisd commented on src/controls/keyboard.py:
```diff
@@ -0,0 +1,93 @@
+
+    def on_message_front_sensor(self, client, userdata, msg):
+        print(f"ultrasonic - {msg.payload}")
+        if int(msg.payload) > 200 and self.up is True:
+            MQTT_VELOCITY_FORWARD.publish(TOPIC.throttle_forward, IDLE)
+            self.crash_expected = True
+        else:
+            self.crash_expected = False
+
+    def on_message_infrared_front(self, client, userdata, msg):
+        print(f"infrared - {msg.payload}")
+        if int(msg.payload) > 0 and self.up is True:
+            MQTT_VELOCITY_FORWARD.publish(TOPIC.throttle_forward, IDLE)
+            self.crash_expected = True
+        else:
+            self.crash_expected = False
```
I don't get this, if you receive a distance **more** than 2 meters then you stop and when you receive an infrared distance that's not 0 you throttle?

VernitaJ replied:
The simple implementation we were going for was that if the sensor read anything at all, other than zero, we would stop it running. 
Instead of this I have adapted the smartcar.ino file to apply the obstacle avoidance, and incorporated logic that works off a range instead. @DrakeAxelrod and @MrJonte are adding some more logic now and will push the code a bit later today.

platisd reviewed (commented):
`Keyboard` is a very strange choice -architecturally- to have your driving logic :}

drakeaxelrod committed:
Update obstacle avoidance

Moved logic for obstacle avoidance from python to arduino file.
This creates a more reliable and robust safety mechanism. Minor
refactoring of both the python code and the arduino file.
